src/ledger/cleanup_service.zig

```diff
@@ -33,19 +33,20 @@ const LOOP_LIMITER = Duration.fromMillis(DEFAULT_CLEANUP_SLOT_INTERVAL * DEFAULT
 pub fn run(
     logger: sig.trace.Logger,
     blockstore_reader: *BlockstoreReader,
-    db: *BlockstoreDB,
+    db: BlockstoreDB,
     lowest_cleanup_slot: *sig.sync.RwMux(Slot),
     max_ledger_shreds: u64,
     exit: *AtomicBool,
 ) !void {
     var last_purge_slot: Slot = 0;
+    var mut_db = db;
```

Comment:
The database is passed in here by value, which is fine (at least for rocksdb). You can just copy it and it will keep working fine with other copies floating around, since it essentially just contains pointers. But then later on, a mutable pointer to the database is needed, since that's how the database methods are written. So it needs to be converted to a var somewhere along the line.

We could try to move to exclusively using a mutable pointer to the database. Just allocate the database struct in one place and pass around the pointer everywhere. One step in this direction would be to change this function to accept it by pointer, and then in its calling scope you convert the const to a var.

But I don't think I like either of those approaches because it feels wrong for the database methods to require mutable pointers to self when self is essentially just a pointer already. I think I'd prefer to conceptualize the database itself as inherently being a pointer, and all of its methods can take self by value (or *const) instead. This also has the benefit of avoiding footguns if we decide to swap out the database implementation with one that is not just a pointer. (I know that sounds counterintuitive, but it centralizes the concern to one place where you would be forced to deal with it, instead of needing to be careful everywhere it is used).

Reply:
okay, if its all pointers that makes sense - i think reading the code it gets confusing though because `BlockstoreDB` doesnt say 'i only contain pointers' just from reading it -- would prefer to stick to *Blockstore because its less confusing and how we do it everywhere else

Reply:
> BlockstoreDB doesnt say 'i only contain pointers' just from reading it

True. It's like a fat pointer taken to the extreme. My main idea is that the only data that should ever be mutated in the database is behind a single pointer field in the struct. The rest is static metadata. So its methods ideally would not require a mutable pointer. Eventually I may make that change, but it's a more significant piece of work than we need right now. It's easier to just change this function to take a mutable pointer.

Reply:
b9deca85dca92001543b28c892137b629379e2dd